proxy/tcp/sni_proxy.go

```diff
@@ -34,27 +36,86 @@ type SNIProxy struct {
 	Noroute metrics.Counter
 }
 
+// Create a buffer large enough to hold the client hello message including
+// the tls record header and the handshake message header.
+// The function requires at least the first 9 bytes of the tls conversation
+// in "data".
+// nil, error is returned if the data does not follow the
+// specification (https://tools.ietf.org/html/rfc5246) or if the client hello
+// is fragmented over multiple records.
+func createClientHelloBuffer(data []byte) ([]byte, error) {
+	// TLS record header
+	// -----------------
+	// byte   0: rec type (should be 0x16 == Handshake)
+	// byte 1-2: version (should be 0x3000 < v < 0x3003)
+	// byte 3-4: rec len
+	if len(data) < 9 {
+		return nil, errors.New("At least 9 bytes required to determine client hello length")
+	}
+
+	if data[0] != 0x16 {
+		return nil, errors.New("Not a TLS handshake")
+	}
+
+	recordLength := int(data[3])<<8 | int(data[4])
+	if recordLength <= 0 || recordLength > 16384 {
+		return nil, errors.New("Invalid TLS record length")
+	}
+
+	// Handshake record header
+	// -----------------------
+	// byte   5: hs msg type (should be 0x01 == client_hello)
+	// byte 6-8: hs msg len
+	if data[5] != 0x01 {
+		return nil, errors.New("Not a client hello")
+	}
+
+	handshakeLength := int(data[6])<<16 | int(data[7])<<8 | int(data[8])
+	if handshakeLength <= 0 || handshakeLength > recordLength-4 {
+		return nil, errors.New("Invalid client hello length (fragmentation not implemented)")
+	}
+
+	return make([]byte, handshakeLength+9), nil //9 for the header bytes
+}
+
 func (p *SNIProxy) ServeTCP(in net.Conn) error {
 	defer in.Close()
 
 	if p.Conn != nil {
 		p.Conn.Inc(1)
 	}
 
-	// capture client hello
-	data := make([]byte, 1024)
-	n, err := in.Read(data)
+	tlsReader := bufio.NewReader(in)
+	data, err := tlsReader.Peek(9)
+	if err != nil {
+		log.Print("[DEBUG] tcp+sni: TLS handshake failed (failed to peek data)")
+		if p.ConnFail != nil {
+			p.ConnFail.Inc(1)
+		}
+		return err
+	}
+
+	tlsData, err := createClientHelloBuffer(data)
```

Comment:
I think I'd prefer if this function would only determine the size of the buffer and not allocate it. Also, keep it in the `tls_clienthello.go` since this is where the parsing logic for the TLS ClientHello header is. Also, I'd keep `data` instead of `tlsData` since there is no difference.